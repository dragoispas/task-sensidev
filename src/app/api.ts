
const BASE_URL = 'https://rickandmortyapi.com/api';

export type CharacterFilters = { name?: string; status?: string; page?: number };

export async function getAllCharacters({ name, status, page }: CharacterFilters = {}) {
    const params = new URLSearchParams({
        ...(name ? { name } : {}),
        ...(status ? { status } : {}),
        ...(page ? { page: page.toString() } : {}),
    });

    const queryString = Array.from(params).length > 0 ? `?${params}` : '';

    try {
        const response = await fetch(`${BASE_URL}/character${queryString}`, {
            method: 'GET',
        });

        if (response.status === 404) {
            return [];
        }

        return response.json();
    } catch (err) {
        return null;
    }
}

//page=3&
export async function getCharacter(characterId: number) {
    try {
        const response = await fetch(`${BASE_URL}/character/${characterId}`, {
            method: 'GET',
        });

        return response.json();
    } catch (err) {
        return null;
    }
}

export async function getEpisodes(ids: number[]) {
    try {

        const response = await fetch(`${BASE_URL}/episode/${ids.join(',')}`, {
            method: 'GET',
        });

        let result = await response.json();

        if (!Array.isArray(result)) {
            result = [result];
        }

        return result;
    } catch (err) {
        return [];
    }
}

