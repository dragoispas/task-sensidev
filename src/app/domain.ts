export type Character = {
    id: number;
    name: string;
    gender: string;
    status: string;
    image: string;
    location: {
        name: string;
    };
    species: string;
    origin: {
        name: string;
    };
    episode: string[];
};

export type Episode = {
    id: number;
    name: string;
    episode: string;
}
