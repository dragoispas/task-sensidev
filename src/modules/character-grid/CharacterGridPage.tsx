export const CharacterGridPage = () => {
    return <div></div>;
}