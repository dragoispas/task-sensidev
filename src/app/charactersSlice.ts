import { createSlice } from '@reduxjs/toolkit';
import type { PayloadAction } from '@reduxjs/toolkit';
import { Character } from './domain';
import { CharacterFilters, getAllCharacters } from './api';
import { AppDispatch } from './store';

export type CharacterMap = {
  [id: number]: Character
}

export interface CharactersState {
  charactersIds: number[],
  charactersMap: CharacterMap,
  page: number,
  maxPage: number,
  errorMessage: string | null
}

const initialState: CharactersState = {
  charactersIds: [],
  charactersMap: {},
  page: 1,
  maxPage: 1,
  errorMessage: null
};

export const charactersSlice = createSlice({
  name: 'characters',
  initialState,
  reducers: {
    setCharacters: (state, action: PayloadAction<Character[]>) => {
      state.charactersIds = action.payload.map(character => character.id);

      const charactersMap: CharacterMap = {};
      action.payload.forEach(character => charactersMap[character.id] = character);
      state.charactersMap = charactersMap;
    },
    setCharactersMap: (state, action: PayloadAction<CharacterMap>) => {
      state.charactersMap = action.payload;
    },
    addCharacterToMap: (state, action: PayloadAction<Character>) => {
      const character = action.payload;
      state.charactersMap[character.id] = character;
    },
    appendCharacters: (state, action: PayloadAction<Character[]>) => {
      state.charactersIds = state.charactersIds.concat(action.payload.map(character => character.id));
      action.payload.forEach(character => state.charactersMap[character.id] = character);
    },
    setPage: (state, action: PayloadAction<number>) => {
      state.page = action.payload;
    },
    incrementPage: (state, action: PayloadAction<void>) => {
      state.page += 1;
    },
    setMaxPage: (state, action: PayloadAction<number>) => {
      state.maxPage = action.payload;
    },
    setErrorMessage: (state, action: PayloadAction<string | null>) => {
      state.errorMessage = action.payload;
    }
  }
});

export const { setCharacters, addCharacterToMap, appendCharacters, setPage, incrementPage, setMaxPage, setErrorMessage } = charactersSlice.actions;

export const retrieveAllCharacters = (filters?: CharacterFilters) => async (dispatch: AppDispatch) => {
  try {
    dispatch(setErrorMessage(null));

    const response = await getAllCharacters(filters);

    if (response === null) {
      dispatch(setErrorMessage('An error has occurred while contacting the API.'));
    }

    dispatch(setCharacters(response?.results ?? []));
    dispatch(setMaxPage(response?.info?.pages ?? 1));
    dispatch(setPage(1));
  } catch (error) {
    console.log(error);
  }
}

export const retrieveMoreCharacters = (filters?: CharacterFilters) => async (dispatch: AppDispatch, getState: any) => {
  const { characters: {  maxPage }} = getState();
  if (filters?.page && filters?.page > maxPage) {
    return;
  }

  dispatch(incrementPage());
  const response = await getAllCharacters(filters);
  dispatch(appendCharacters(response?.results ?? []));
}

export default charactersSlice.reducer;