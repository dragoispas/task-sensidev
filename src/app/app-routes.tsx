import { Route, Routes } from 'react-router-dom';
import { ErrorPage } from '../pages/ErrorPage';
import { CharacterGridPage } from '../pages/CharacterGridPage';
import { CharacterDetailsPage } from '../pages/CharacterDetailsPage';

export default function AppRoutes() {
  return (
    <Routes>
      <Route path="/character/:id" element={<CharacterDetailsPage />} />
      <Route path="/" element={<CharacterGridPage />} />
      <Route path="*" element={<ErrorPage />} />
    </Routes>
  );
}