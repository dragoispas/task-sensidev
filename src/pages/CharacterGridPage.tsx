import styled from "@emotion/styled";
import { useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { retrieveAllCharacters, retrieveMoreCharacters } from "../app/charactersSlice";
import { AppDispatch, RootState } from "../app/store";
import { useDebounce } from "../app/debounce";
import { CharacterCard } from "../components/CharacterCard";
import { SearchBar } from "../components/SearchBar";
import { Button, Stack } from "../components/common";

const CharacterGrid = styled.div`
    display: flex;
    flex-wrap: wrap;
    align-items: center;
    justify-content: center;
    max-width: 1560px;

    margin: 20px auto;
`;

export const CharacterGridPage = () => {
    const dispatch = useDispatch();
    const { characters, page, maxPage } = useSelector((state: RootState) => state.characters);

    const [searchInput, setSearchInput] = useState<string>('');
    const [status, setStatus] = useState<string>('');

    const debouncedSearchInput = useDebounce(searchInput, 300);
    
    useEffect(() => {
        (dispatch as AppDispatch)(retrieveAllCharacters({ name: debouncedSearchInput, status: status }));
    }, [debouncedSearchInput, status, dispatch]);

    return (
        <Stack justifyContent="center" alignItems="center">
            <img style={{margin: '10px auto'}} src="images/Rick&Morty.png" alt='.'></img>
            <SearchBar searchInput={searchInput} setSearchInput={setSearchInput} status={status} setStatus={setStatus}/>
            <CharacterGrid>
                {
                    characters ? (
                        characters.map((character) => (
                            <CharacterCard key={character.id} id={character.id} name={character.name.toUpperCase()} status={character.status} image={character.image}/>
                        ))
                    ) : null
                }
            </CharacterGrid>
            { 
                page === maxPage || characters.length === 0 ? null : (
                    <Button onClick={() => {(dispatch as AppDispatch)(retrieveMoreCharacters({ name: debouncedSearchInput, status: status, page: page + 1 }));}}>LOAD MORE</Button>
                )
            }
        </Stack>
    );
}