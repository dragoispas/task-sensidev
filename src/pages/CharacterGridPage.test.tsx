import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { store } from '../app/store';
import { CharacterGridPage } from './CharacterGridPage';

test('Renders the grid page and matches snapshot', () => {
  const tree = renderer.create(
    <Provider store={store}>
      <CharacterGridPage />
    </Provider>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
