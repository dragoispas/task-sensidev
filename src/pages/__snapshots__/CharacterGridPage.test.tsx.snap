// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders the grid page and matches snapshot 1`] = `
<div
  className="css-1vjcx31"
>
  <img
    alt="."
    src="images/Rick&Morty.png"
    style={
      Object {
        "margin": "10px auto",
      }
    }
  />
  <div
    className="css-6v0k7w"
  >
    <input
      className="css-6qeqrq"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value=""
    />
    <div
      className="css-17vjr3m"
      color="#abdb85"
      height="36px"
      width="2px"
    />
    <select
      className="css-1l2janw"
      onChange={[Function]}
      value=""
    >
      <option
        value=""
      >
        all
      </option>
      <option
        value="alive"
      >
        alive
      </option>
      <option
        value="dead"
      >
        dead
      </option>
      <option
        value="unknown"
      >
        unknown
      </option>
    </select>
  </div>
  <div
    className="css-kj7lrm"
  />
</div>
`;
